Orange/data/table.py

```diff
@@ -539,225 +590,65 @@ def from_url(cls, url):
-            r_metas = [attributes[i]
-                       for i, col in enumerate(col_indices) if col < 0]
-            domain = Domain(r_attrs, r_classes, r_metas)
-        else:
-            domain = self.domain
-        return self.from_table(domain, self, row_idx)
+    def add_column(self, col_class, column_types, columns):
+        # TODO: override __setitem__ for strings, autodetermine variable type, default role X
+        """
+        Add one or multiple columns to the table.
+        Column types and column values must be of the same length.
+        If only a single value for a given column exists, the value is used in all rows (broadcasting).
+        :param col_class: Specifies the target column class (either x, y, or meta, see Table.ColumnClass).
+        :param column_types: The column type(s), in the form of Variables.
+        :param columns: The column values. Allows broadcasting single values.
+        """
+        column_types, columns = self._transform_types_columns(column_types, columns)
+        for t, c in zip(column_types, columns):
+            self._column_class_map[t] = Table.ColumnClass.get(col_class)
+            self[t] = c
 
     def __setitem__(self, key, value):
-        if not self._check_all_dense():
-            raise ValueError(
-                "Assignment to rows of sparse data is not supported")
-        if not isinstance(key, tuple):
-            if isinstance(value, Real):
-                self.X[key, :] = value
-                return
-            self._set_row(value, key)
-            return
-
-        if len(key) != 2:
-            raise IndexError("Table indices must be one- or two-dimensional")
-        row_idx, col_idx = key
-
-        # single row
-        if isinstance(row_idx, Integral):
-            if isinstance(col_idx, slice):
-                col_idx = range(*slice.indices(col_idx, self.X.shape[1]))
-            if not isinstance(col_idx, str) and isinstance(col_idx, Iterable):
-                col_idx = list(col_idx)
-            if not isinstance(col_idx, str) and isinstance(col_idx, Sized):
-                if isinstance(value, (Sequence, np.ndarray)):
-                    values = value
-                elif isinstance(value, Iterable):
-                    values = list(value)
-                else:
-                    raise TypeError("Setting multiple values requires a "
-                                    "sequence or numpy array")
-                if len(values) != len(col_idx):
-                    raise ValueError("Invalid number of values")
+        # we only override this for certain types:
+        #  - Variables
+        #  - plain strings (their Variable is constructed on the fly)
+        # if the column is already in the table, we aren't adding new columns
+        # otherwise we pass this along to the parent
+        if not (isinstance(key, Variable) or isinstance(key, str)) or key in self.columns:
+            super(Table, self).__setitem__(key, value)
+
```

Comment:
Yes. Exactly what `self.insert(len(self.columns), key, value)` does (or should do).